### ISSUE TITLE
ERRO You are not allowed to use this program, Caused by: N/A

### ISSUE THREAD

**Issue by ratijas** [Frederick888/git-credential-keepassxc#20]
## Summary

`git-credential-keepassxc` fails to communicate with KeePassXC.

```sh
❯ git-credential-keepassxc configure
Dec 02 23:28:37.623 ERRO You are not allowed to use this program, Caused by: N/A
```

## Steps to reproduce

1. Database is unlocked (otherwise a different error is returned),
2. Run `git-credential-keepassxc configure`,
3. KeePassXC prompt for a connection name, 
4. Type in any name, press "Save and allow access" button.

## Expected result

Pairing complete, connection information stored.

## Observed result

An error is reported in terminal. Presumably because KeePassXC refuses to communicate with `git-credential-keepassxc`.

## Context

KeePassXC: 2.6.2-1
git-credential-keepassxc: 0.4.3, features: all
rust: 1.48.0/stable
OS: Arch Linux

I figured, my browser integration settings might be relevant, so here they are.

![browser integration settings - general](https://user-images.githubusercontent.com/6737986/100928732-a5504600-34f7-11eb-8356-ef0b4d30deb6.png)
![browser integration settings - advanced](https://user-images.githubusercontent.com/6737986/100928893-e21c3d00-34f7-11eb-82d7-79a47cda936b.png)

**Comment by ratijas** [comment]
Oh well, I need new glasses.

> Note: If you've enabled strict-caller, you must add caller profiles before configuring databases, otherwise you won't be able to run git-credential-keepassxc afterwards.

**Comment by ratijas** [comment]
On a second thought, in spirit with rust ecosystem, error message can be improved. Something like: "Have you forgotten to run `git-credential-keepassxc caller add` first?" (not sure about english grammar) would have prevented this issue from even being opened in the first place.  README is good, but clear actionable runtime messages are important too. What do you think, @Frederick888?

**Comment by Frederick888** [comment]
@ratijas Sure thing. I'll print such a message when `strict-caller` is enabled while there're no callers defined.

The 'symbol link caller path' feature currently I've got on `master` branch needs to be guarded by a configuration, so this will be squeezed into next release.

I'll reopen this ticket for tracking. Thanks for the suggestion.

**Comment by ratijas** [comment]
Also, is it really needed to log date&amp;time to the console while in interactive mode? (Interactive mode usually simply means that stdout is connected to a tty.)

**Comment by Frederick888** [comment]
It's the default format from the logging dependency I use (slog, slog-term) and I feel totally fine with it.

Logs are printed to stderr I believe. You can easily strip out the date & time using `sed` or something.

**Comment by ratijas** [comment]
I mean, it just looks too formal for a utility like this. Like if it was a freaking nginx or something.

Also, this is not very user-friendly message:

```
$ git send-email *.patch
...

Send this email? ([y]es|[n]o|[e]dit|[q]uit|[a]ll): 
Dec 03 01:53:50.938 ERRO Request get-logins failed, No logins found (code: 15), Caused by: N/A
Password for 'smtp://me@ratijas.tk@smtp.yandex.ru:465': 
```

I kinda understand that it hasn't found credentials. Obviously, because it's the first time I'm using this helper with git. But come on, that's no reason to shout in my face like that. That's not even an error — just credentials do not exist yet.

Works smoothly on a second send though. Cool. Might as well be the best thing ever since sliced bread. :slightly_smiling_face:

**Comment by Frederick888** [comment]
Haha. But honestly I think it's something that may vary from person to person. As I mentioned above, you can have a simple wrapper:

```sh
#!/usr/bin/env bash

git-credential-keepassxc "$@" 2> >(sed 's/.*\(\(ERRO\|INFO\|WARN\).*\)/\1/')
```

**Comment by ratijas** [comment]
> As I mentioned above, you can have a simple wrapper

Nah... of course I'm too lazy for that.

I still stand by the point that missing credentials is not an error, but rather an informative message or (at worst) a warning. Consider this: if you were a KeePassXC browser plugin, you wouldn't be flooding UI with errors screaming "hey, there is a form, but I can't fill it in!" — just because user visits some site for the first time, or would you?

**Comment by ratijas** [comment]
> Frederick888 closed this in 8365809 11 minutes ago

Thanks, Frederick!  I appreciate that.
